File: CrossSetaDeduplicator/src/CrossSetaWeb/CrossSetaLogic/Services/ValidationProgressService.vb
Imports System.Collections.Concurrent
Imports CrossSetaLogic.Models

Namespace Services
    Public Class ValidationProgressService
        Implements IValidationProgressService

        Private Shared ReadOnly _jobs As New ConcurrentDictionary(Of String, ValidationProgress)()

        Public Sub StartValidation(jobId As String) Implements IValidationProgressService.StartValidation
            _jobs(jobId) = New ValidationProgress With {
                .Total = 0,
                .Processed = 0,
                .Status = "Starting...",
                .IsComplete = False
            }
        End Sub

        Public Sub UpdateProgress(jobId As String, processed As Integer, total As Integer, status As String) Implements IValidationProgressService.UpdateProgress
            Dim progress As ValidationProgress = Nothing
            If _jobs.TryGetValue(jobId, progress) Then
                progress.Processed = processed
                progress.Total = total
                progress.Status = status
            End If
        End Sub

        Public Sub CompleteValidation(jobId As String, result As DatabaseValidationResult) Implements IValidationProgressService.CompleteValidation
            Dim progress As ValidationProgress = Nothing
            If _jobs.TryGetValue(jobId, progress) Then
                progress.IsComplete = True
                progress.Status = "Completed"
                progress.Result = result
                progress.Processed = progress.Total ' Ensure 100%
            End If
        End Sub

        Public Function GetProgress(jobId As String) As ValidationProgress Implements IValidationProgressService.GetProgress
            Dim progress As ValidationProgress = Nothing
            If _jobs.TryGetValue(jobId, progress) Then
                Return progress
            End If
            Return Nothing
        End Function
    End Class
End Namespace

File: CrossSetaDeduplicator/src/CrossSetaWeb/Services/ValidationProgressService.cs
using System;
using System.Collections.Concurrent;
using System.Linq;

namespace CrossSetaWeb.Services
{
    public class ValidationProgress
    {
        public int Total { get; set; }
        public int Processed { get; set; }
        public int Percentage => Total == 0 ? 0 : (int)((double)Processed / Total * 100);
        public string Status { get; set; }
        public bool IsComplete { get; set; }
        public DatabaseValidationResult Result { get; set; }
    }

    public interface IValidationProgressService
    {
        void StartValidation(string jobId);
        void UpdateProgress(string jobId, int processed, int total, string status);
        void CompleteValidation(string jobId, DatabaseValidationResult result);
        ValidationProgress GetProgress(string jobId);
    }

    public class ValidationProgressService : IValidationProgressService
    {
        private static readonly ConcurrentDictionary<string, ValidationProgress> _jobs = new ConcurrentDictionary<string, ValidationProgress>();

        public void StartValidation(string jobId)
        {
            _jobs[jobId] = new ValidationProgress 
            { 
                Total = 0, 
                Processed = 0, 
                Status = "Starting...", 
                IsComplete = false 
            };
        }

        public void UpdateProgress(string jobId, int processed, int total, string status)
        {
            if (_jobs.TryGetValue(jobId, out var progress))
            {
                progress.Processed = processed;
                progress.Total = total;
                progress.Status = status;
            }
        }

        public void CompleteValidation(string jobId, DatabaseValidationResult result)
        {
            if (_jobs.TryGetValue(jobId, out var progress))
            {
                progress.IsComplete = true;
                progress.Status = "Completed";
                progress.Result = result;
                progress.Processed = progress.Total; // Ensure 100%
            }
        }

        public ValidationProgress GetProgress(string jobId)
        {
            if (_jobs.TryGetValue(jobId, out var progress))
            {
                return progress;
            }
            return null;
        }
    }
}
